feat: add usage-insight on the desktop-sidebar

revogabe opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

Fixes # (issue)

*If there is not an issue for this, please create one first. This is used to tracking purposes and also helps use understand why this PR exists*

<!-- If there isn't an issue for this PR, please re-review our Contributing Guide and create an issue -->

## Type of change

<!-- Please mark the relevant points by using [x] -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [X] Enhancement (small improvements)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

1. **Upgrade Card Display**  
   - Log in to the dashboard and verify that the Upgrade Card appears for users with request-based plans.  

2. **Usage Data Accuracy**  
   - Compare the token usage displayed on the card with actual API request logs.  
   - Confirm that the progress bar updates dynamically as requests are made.  

3. **Upgrade Flow**  
   - Click the **Upgrade** button and ensure it correctly redirects to the billing page.  
   - Test the redirection on different devices and browsers to confirm consistent behavior.  

4. **Edge Cases**  
   - Verify how the card behaves when a user reaches **90%** and **100%** of their limit.  
   - Check if the UI remains responsive when displaying large token usage numbers.  

5. **Performance & UX**  
   - Ensure the Upgrade Card does not negatively impact dashboard load times.  
   - Test different screen sizes to confirm a responsive design.  

## Checklist

<!-- We're starting to get more and more contributions. Please help us making this efficient for all of us and go through this checklist. Please tick off what you did  -->

### Required

- [X] Filled out the "How to test" section in this PR
- [X] Read [Contributing Guide](./CONTRIBUTING.md)
- [X] Self-reviewed my own code
- [X] Commented on my code in hard-to-understand areas
- [X] Ran `pnpm build`
- [X] Ran `pnpm fmt`
- [X] Checked for warnings, there are none
- [X] Removed all `console.logs`
- [X] Merged the latest changes from main onto my branch with `git pull origin main`
- [X] My changes don't cause any responsiveness issues

### Appreciated

- [X] If a UI change was made: Added a screen recording or screenshots to this PR
- [X] Updated the Unkey Docs if changes were necessary

https://github.com/user-attachments/assets/498b0b1b-7e57-40b9-9c47-223ef7d06c9a

revogabe commented:
Done :wave:

ogzhanolguncu commented on apps/dashboard/app/(app)/settings/billing/components/usage-insights.tsx:
```diff
@@ -0,0 +1,233 @@
+        <ProgressCircle
+          value={item?.current ?? current}
+          max={item?.max ?? max}
+          color={color ?? "#f76e19"}
```
Please, let's use colors from our palette. For this one you can probably use `orange`.

ogzhanolguncu commented on apps/dashboard/app/(app)/settings/billing/components/usage-insights.tsx:
```diff
@@ -0,0 +1,233 @@
+export const Item = React.forwardRef<PrimitiveDivElement, UsageItemProps>(
+  (props, ref) => {
+    const { item, title, description, color, className, ...itemProps } = props
+    const { format } = Intl.NumberFormat(undefined, { notation: "compact" })
```
Use `formatNumber` from the utils in `lib/fmt`.

ogzhanolguncu commented on apps/dashboard/lib/mouse.tsx:
Since this is a hook let's call it `use-mouse`. And, I love to move this to hooks dir.

ogzhanolguncu reviewed (commented):
I'm not sure why I don't see the use of `<Particles />` component and `useMouse` hook. This is how it looks locally.

<img width="311" alt="Screenshot 2025-03-18 at 13 08 27" src="https://github.com/user-attachments/assets/c482a6f1-4a67-4b3d-b555-a282f9f63791" />

Maybe its a weird fork issue. But on preview everything works great 🤔

chronark commented on apps/dashboard/components/ui/particles.tsx:
we don't want to use this in the dashboard. let's remove this file

revogabe commented:
it's make sense ? cc: @ogzhanolguncu 

![image](https://github.com/user-attachments/assets/399d2f97-cb4c-46e4-8d34-2fc46f7a8dc0)

chronark commented on apps/dashboard/hooks/use-mouse.tsx:
do we need this?

revogabe replied:
As @ogzhanolguncu said:

_"Since this is a hook, let's call it useMouse. And I'd love to move this to the hooks directory."_

I thought he wanted to keep this hook in the hooks folder, but we don't really need to.

chronark replied:
if we're not using it, then let's not add it at all
afaik it was only needed for the particles

ogzhanolguncu commented:
I guess it's just a leftover from particles version

revogabe commented:
do we have any blockers? cc: @ogzhanolguncu

perkinsjr reviewed (changes requested):
I am seeing issues with this. For example


This one shows incorrectly, shows Pro Extra but 150k requests.
<img width="1822" alt="Screenshot 2025-03-19 at 6 32 17 PM" src="https://github.com/user-attachments/assets/29fd39fd-81ed-4bea-b488-e598b1bbb488" />

Then I created a new workspace added a credit card and added pro tier 

Now it shows Free tier but the right amount

<img width="1710" alt="Screenshot 2025-03-19 at 6 34 37 PM" src="https://github.com/user-attachments/assets/5c3d875d-ff7b-4c17-8be2-6ddea8e509ae" />

revogabe commented:
Investigating 🤔

revogabe commented:
This issue is probably related to what I discussed with @chronark . We have a bug in the request for workspace?.quota, specifically when we are in a personal workspace. However, when I create a new workspace, I verified that the quota is created normally. So, it's likely a bug in the quota creation when creating an account.

Regarding the plan being displayed incorrectly while in the workspace, I haven't figured out the cause yet. I'm working on a solution for these issues. cc: @perkinsjr

revogabe commented on apps/dashboard/app/(app)/desktop-sidebar.tsx:
```diff
@@ -94,7 +106,17 @@ export const DesktopSidebar: React.FC<Props> = ({ workspace, className }) => {
+      <div className="bg-[inherit] min-w-full [flex:0_0_56px] -mx-2 sticky flex flex-col justify-end items-start pb-3 gap-3">
+        <UsageInsight.Root
+          isLoading={usageQuery.isFetching}
+          plan={workspace.tier ?? "Free"}
```
i'm using workspace.tier cc: @chronark

revogabe replied:
about the animation, it really makes sense, I will remove the revalidation cache and just put a skeleton, about the refetch do you think we should put a refetchInterval or just when changing pages or leaving and entering it?

chronark replied:
I see, well it looked like you used plan, cause you named the variables `plan`

ie: `<h2 className="text-gray-12 text-lg capitalize">{dataToUse.plan}</h2>
`

chronark replied:
I don't think we need periodic refetches here

revogabe commented:
I followed the step-by-step process from account creation to workspace creation. Note that I set a random value for the prop to check if it's not returning the quota correctly.

https://github.com/user-attachments/assets/e4a279ed-ebae-4fe2-86f1-8a713084c37b

I also checked inside the database and noticed that for personal workspaces, no quotas are being created, only for additional workspaces.

https://github.com/user-attachments/assets/a8f2ee06-2cc9-4ba3-bb10-2bd4a6b6ed69

ogzhanolguncu commented:
<img width="595" alt="image" src="https://github.com/user-attachments/assets/6f654245-caec-49c3-963f-9814d908a48a" />
Can we also make sure skeleton colors are the same as the chart?

chronark commented:
hey @revogabe thanks for finding that bug
it's fixed in #3007, just do a pull from main again

revogabe commented:
- Component adapted for the new sidebar  
- Simple loading and animation using the skeleton component  
- Removed refetch interval  
- Upgrade button only visible above 90%  
- Removed cached data causing bugs in the workspace tier  

Needs to be tested:  
- When upgrading to the Pro plan, check if everything is working correctly with the title and the maximum number of requests.

mcstepp commented:
The stripe variable keys are in apps/dashboard/lib/env.ts. You'll have to supply your own stripe variable values though. We can't give out Unkey's values.

mcstepp commented:
We also fixed it so that quotas table is generated for first-time users' personal workspaces, so after pulling from main, when you clear your database (bad data), it should include the quotas table now.

chronark commented:
yeah this should be fixed,
regardless I can take a look in the morning, but as meg said, unless contributors want to setup their own stripe, they can’t test it
